src/main/java/com/ge/predix/uaa/token/lib/FastTokenServices.java

```diff
@@ -79,12 +83,16 @@
 
     private List<String> trustedIssuers;
 
-    private final Map<String, SignatureVerifier> tokenKeys = new HashMap<>();
+    private long timeToLiveMillis = EXPIRATION_TIME_IN_MILLIS;
+
+    private SignatureVerifierProvider signatureVerifierProvider = new NetworkOutageTolerantSignatureVerifierProvider();
```

Comment:
@RoopGuron  I haven't reviewed details of NetworkOutage...Provider - is the main purpose of this class to ensure that it expires the verifying key based on a TTL?  In that case, you can simply change the tokenKeys map above to a `PassiveExpiringMap` instead of `HashMap`.  Will that suffice? or I am missing something?

Reply:
Apart from TTL, UAA unavailability has to be taken into consideration. It goes like this, even though the **issuer** **tokenKey** has expired if the UAA is unavailable at that instant, the **HashMap** needs to retain the **tokenKey**.

Reply:
I disagree. If the validator cannot fetch the key after cache has expired, authentication should fail. This will make sure that if the key has been revoked, it will never be used after the configured TTL has elapsed.

Reply:
Even CRL for digital certificates has a lag of 24 hours. I guess the primary driver for this is degraded experience for uaa users.

Reply:
Do you agree that failing the token validation when UAA is down, is more secure?  Not sure what you mean by backward compatibility when fixing a security vulnerability. Incompatibility here is desired :)

Reply:
These are the requirements we have to implement.
we can have a separate discussion with PMs on requirements.